prog/vm/nexus.rb

```diff
@@ -75,23 +82,30 @@ def self.assemble(public_key, project_id, name: nil, size: "standard-2",
 
       vm.associate_with_project(project)
 
-      if storage_encrypted
-        key_encryption_key = StorageKeyEncryptionKey.create_with_id(
-          algorithm: key_wrapping_algorithm,
-          key: Base64.encode64(key_wrapping_key),
-          init_vector: Base64.encode64(key_wrapping_iv),
-          auth_data: "#{vm.inhost_name}_0"
+      storage_volumes.each_with_index do |volume, disk_index|
+        if volume[:encrypted]
+          key_wrapping_algorithm = "aes-256-gcm"
+          cipher = OpenSSL::Cipher.new(key_wrapping_algorithm)
+          key_wrapping_key = cipher.random_key
+          key_wrapping_iv = cipher.random_iv
+
+          key_encryption_key = StorageKeyEncryptionKey.create_with_id(
+            algorithm: key_wrapping_algorithm,
+            key: Base64.encode64(key_wrapping_key),
+            init_vector: Base64.encode64(key_wrapping_iv),
+            auth_data: "#{vm.inhost_name}_#{disk_index}"
+          )
+        end
+
+        VmStorageVolume.create_with_id(
+          vm_id: vm.id,
+          boot: disk_index == 0,
+          size_gib: volume[:size_gib],
+          disk_index: disk_index,
+          key_encryption_key_1_id: volume[:encrypted] ? key_encryption_key.id : nil
```

Comment:
nit: we might not need to check `volume[:encrypted]` twice
```suggestion
        key_encryption_key = if volume[:encrypted]
          key_wrapping_algorithm = "aes-256-gcm"
          cipher = OpenSSL::Cipher.new(key_wrapping_algorithm)
          key_wrapping_key = cipher.random_key
          key_wrapping_iv = cipher.random_iv

          StorageKeyEncryptionKey.create_with_id(
            algorithm: key_wrapping_algorithm,
            key: Base64.encode64(key_wrapping_key),
            init_vector: Base64.encode64(key_wrapping_iv),
            auth_data: "#{vm.inhost_name}_#{disk_index}"
          )
        end

        VmStorageVolume.create_with_id(
          vm_id: vm.id,
          boot: disk_index == 0,
          size_gib: volume[:size_gib],
          disk_index: disk_index,
          key_encryption_key_1_id: key_encryption_key&.id
```